### PR TITLE
Add ChannelReader.ReadAllAsync to System.Threading.Channels

### PR DESCRIPTION
**Pull request by stephentoub**
Fixes https://github.com/dotnet/corefx/issues/32742

cc: @tarekgh, @davidfowl, @jcouv

### CODE REVIEW

File: src/System.Threading.Channels/src/System/Threading/Channels/ChannelReader.netcoreapp.cs [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+
+            public void MoveNext()
+            {
+                try
+                {
+                    bool waitToReadResult = false;
+                    switch ((State)_state)
+                    {
+                        case State.OuterLoop:
+                            ValueTask<bool> waitToReadTask = _reader.WaitToReadAsync(_cancellationToken);
```
Comment (tarekgh) [inline src/System.Threading.Channels/src/System/Threading/Channels/ChannelReader.netcoreapp.cs]: >_reader.WaitToReadAsync(_cancellationToken) [](start = 61, length = 43)

do we need to have .ConfigureAwait(false) here?

File: src/System.Threading.Channels/src/System/Threading/Channels/ChannelReader.netcoreapp.cs [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+                                waitToReadResult = waitToReadTask.GetAwaiter().GetResult();
+                                _state = (int)State.HaveWaitToReadResult;
+                                goto case State.HaveWaitToReadResult;
+                            }
+
+                            // WaitToReadAsync wasn't yet complete.  Mark that the promise
+                            // is being used for the result, set where we should return to when it completes,
+                            // store the awaiter, and hook up the continuation.
+                            _usePromiseForResult = true;
+                            _state = 1;
```
Comment (tarekgh) [inline src/System.Threading.Channels/src/System/Threading/Channels/ChannelReader.netcoreapp.cs]: > 1; [](start = 36, length = 3)

would it be better to use State.FinishingAwait?

**Comment by tarekgh** [comment]
Somehow looks the UWP test run failed for unknown reason

https://mc.dot.net/#/user/dotnet-bot/pr~2Fdotnet~2Fcorefx~2Frefs~2Fpull~2F34756~2Fmerge/test~2Ffunctional~2Fcli~2F/20190122.17/workItem/System.Threading.Channels.Tests/wilogs

```
Executed on a000WBT

C:\dotnetbuild\work\9d63c360-d6ae-496b-b956-996682cbbe16\Work\47a9f602-f241-4383-be3a-d0117dbc0455\Exec>call RunTests.cmd C:\dotnetbuild\work\9d63c360-d6ae-496b-b956-996682cbbe16\Payload 
Using C:\dotnetbuild\work\9d63c360-d6ae-496b-b956-996682cbbe16\Payload as the test runtime folder.
Executing in C:\dotnetbuild\work\9d63c360-d6ae-496b-b956-996682cbbe16\Work\47a9f602-f241-4383-be3a-d0117dbc0455\Exec\ 
----- start 23:43:28.83 ===============  To repro directly: ===================================================== 
pushd C:\dotnetbuild\work\9d63c360-d6ae-496b-b956-996682cbbe16\Work\47a9f602-f241-4383-be3a-d0117dbc0455\Exec\
for /F %f in ('dir /b /s "C:\dotnetbuild\work\9d63c360-d6ae-496b-b956-996682cbbe16\Payload\Runner\"|find /i /v ".pri"') do @(set relPath=%f& set relPath=C:\dotnetbuild\work\9d63c360-d6ae-496b-b956-996682cbbe16\Payload\Runner\=relPathrelPathrelPathrelPath %f 1>NUL))
for /R "C:\dotnetbuild\work\9d63c360-d6ae-496b-b956-996682cbbe16\Payload\UAPLayout\" %I in (*) do @if not exist "%~nxI" (mklink /H "%~nxI" "%I" 1>NUL)
copy /y "C:\dotnetbuild\work\9d63c360-d6ae-496b-b956-996682cbbe16\Payload\UAPLayout\System.Private.CoreLib.dll" System.Private.CoreLib.ni.dll
call "C:\dotnetbuild\work\9d63c360-d6ae-496b-b956-996682cbbe16\Payload\Launcher\WindowsStoreAppLauncher.exe" -install appxmanifest.xml System.Threading.Channels.Tests.dll
echo.
XUnitRunnerUap System.Threading.Channels.Tests.dll -xml "C:\Users\runner\Documents\System.Threading.Channels.Tests.xml" -nologo -notrait category=nonuaptests -notrait category=nonwindowstests -notrait category=IgnoreForCI -notrait category=failing -notrait category=Outerloop
echo.
set ERRORLEVEL=0
call "C:\dotnetbuild\work\9d63c360-d6ae-496b-b956-996682cbbe16\Payload\Launcher\WindowsStoreAppLauncher.exe" -uninstall appxmanifest.xml System.Threading.Channels.Tests.dll
move /Y "C:\Users\runner\Documents\System.Threading.Channels.Tests.xml" .\testResults.xml
popd
===========================================================================================================

C:\dotnetbuild\work\9d63c360-d6ae-496b-b956-996682cbbe16\Work\47a9f602-f241-4383-be3a-d0117dbc0455\Exec>for /F %f in ('dir /b /s "C:\dotnetbuild\work\9d63c360-d6ae-496b-b956-996682cbbe16\Payload\Runner\"|find /i /v ".pri"') do @(set relPath=%f  & set relPath=!relPath:C:\dotnetbuild\work\9d63c360-d6ae-496b-b956-996682cbbe16\Payload\Runner\=!  & if exist %f\nul (if not exist !relPath! mkdir !relPath! )  else (if not exist !relPath! mklink /h !relPath! %f  1>NUL ) ) 

C:\dotnetbuild\work\9d63c360-d6ae-496b-b956-996682cbbe16\Work\47a9f602-f241-4383-be3a-d0117dbc0455\Exec>for /R "C:\dotnetbuild\work\9d63c360-d6ae-496b-b956-996682cbbe16\Payload\UAPLayout\" %I in (*) do @if not exist "%~nxI" (mklink /H "%~nxI" "%I"  1>NUL ) 

C:\dotnetbuild\work\9d63c360-d6ae-496b-b956-996682cbbe16\Work\47a9f602-f241-4383-be3a-d0117dbc0455\Exec>copy /y "C:\dotnetbuild\work\9d63c360-d6ae-496b-b956-996682cbbe16\Payload\UAPLayout\System.Private.CoreLib.dll" System.Private.CoreLib.ni.dll 
        1 file(s) copied.

C:\dotnetbuild\work\9d63c360-d6ae-496b-b956-996682cbbe16\Work\47a9f602-f241-4383-be3a-d0117dbc0455\Exec>call "C:\dotnetbuild\work\9d63c360-d6ae-496b-b956-996682cbbe16\Payload\Launcher\WindowsStoreAppLauncher.exe" -install appxmanifest.xml System.Threading.Channels.Tests.dll 
Got manifest file appxmanifest.xml
Removing any previous installation...
Installing the application...
Package Full Name is 5cd54353-3ed7-4a6e-a72f-db349f28867c_1.0.0.0_x64__v52bfwc2c21ha

SUCCESS
ExitCode 100

C:\dotnetbuild\work\9d63c360-d6ae-496b-b956-996682cbbe16\Work\47a9f602-f241-4383-be3a-d0117dbc0455\Exec>echo.


C:\dotnetbuild\work\9d63c360-d6ae-496b-b956-996682cbbe16\Work\47a9f602-f241-4383-be3a-d0117dbc0455\Exec>XUnitRunnerUap System.Threading.Channels.Tests.dll -xml "C:\Users\runner\Documents\System.Threading.Channels.Tests.xml" -nologo -notrait category=nonuaptests -notrait category=nonwindowstests -notrait category=IgnoreForCI -notrait category=failing -notrait category=Outerloop 
Discovering: System.Threading.Channels.Tests
Discovered:  System.Threading.Channels.Tests
Starting:    System.Threading.Channels.Tests
   System.Threading.Channels.Tests.StressTests.RunInStressMode [SKIP]
      Condition(s) not met: \"IsStressModeEnabled\"
Finished:    System.Threading.Channels.Tests
System.Threading.Channels.Tests  Total: 1020, Errors: 0, Failed: 0, Skipped: 1, Time: 5.6066053

C:\dotnetbuild\work\9d63c360-d6ae-496b-b956-996682cbbe16\Work\47a9f602-f241-4383-be3a-d0117dbc0455\Exec>echo.


C:\dotnetbuild\work\9d63c360-d6ae-496b-b956-996682cbbe16\Work\47a9f602-f241-4383-be3a-d0117dbc0455\Exec>set ERRORLEVEL=0 

C:\dotnetbuild\work\9d63c360-d6ae-496b-b956-996682cbbe16\Work\47a9f602-f241-4383-be3a-d0117dbc0455\Exec>call "C:\dotnetbuild\work\9d63c360-d6ae-496b-b956-996682cbbe16\Payload\Launcher\WindowsStoreAppLauncher.exe" -uninstall appxmanifest.xml System.Threading.Channels.Tests.dll 
Got manifest file appxmanifest.xml
Removing the application...

SUCCESS
ExitCode 100

C:\dotnetbuild\work\9d63c360-d6ae-496b-b956-996682cbbe16\Work\47a9f602-f241-4383-be3a-d0117dbc0455\Exec>move /Y "C:\Users\runner\Documents\System.Threading.Channels.Tests.xml" .\testResults.xml 
The system cannot find the file specified.
----- end 23:43:39.49 ----- exit code 0 ----------------------------------------------------------
2019-01-22 23:43:39,994: ERROR: xunit-reporter(70): main: Unable to report xunit results: no test results xml file found.
```

File: src/System.Threading.Channels/src/System/Threading/Channels/ChannelReader.netcoreapp.cs (hunk 1 above)
Comment (stephentoub) [reply src/System.Threading.Channels/src/System/Threading/Channels/ChannelReader.netcoreapp.cs]: No, it's below when actually creating the awaiter.

File: src/System.Threading.Channels/src/System/Threading/Channels/ChannelReader.netcoreapp.cs (hunk 2 above)
Comment (stephentoub) [reply src/System.Threading.Channels/src/System/Threading/Channels/ChannelReader.netcoreapp.cs]: Oops, yes.

File: src/System.Threading.Channels/src/System/Threading/Channels/ChannelReader.netcoreapp.cs [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+{
+    public abstract partial class ChannelReader<T>
+    {
+        /// <summary>Creates an <see cref="IAsyncEnumerable{T}"/> that enables reading all of the data from the channel.</summary>
+        /// <remarks>
+        /// Each <see cref="IAsyncEnumerator{T}.MoveNextAsync"/> call that returns <c>true</c> will read the next item out of the channel.
+        /// <see cref="IAsyncEnumerator{T}.MoveNextAsync"/> will return false once no more data is or will ever be available to read.
+        /// </remarks>
+        /// <returns></returns>
+        public virtual IAsyncEnumerable<T> ReadAllAsync() => new AsyncEnumerable(this);
```
Comment (davidfowl) [inline src/System.Threading.Channels/src/System/Threading/Channels/ChannelReader.netcoreapp.cs]: Why not use yield?

Comment (davidfowl) [reply src/System.Threading.Channels/src/System/Threading/Channels/ChannelReader.netcoreapp.cs]: Ah

**Pull request merged**